Compilation problem in the authssh2 policy set

This issue probably depends on the openssl version.
I have both openssl 1.0.2.k-3 and 1.1.0.e-1 installed on my systems, but I don't know which one is used for IRATI. Probably the newest.

````
libtool: compile:  g++ -DHAVE_CONFIG_H -I. -I/home/vmaffione/git/nxw/irati/librina/src -DSTUB_API=0 -I/home/vmaffione/git/nxw/irati/librina/include -I/home/vmaffione/git/nxw/irati/librina/src/jsoncpp -I/usr/include/libnl3 -I/usr/include/libnl3 -pthread -Wall -Wformat -Wnon-virtual-dtor -Woverloaded-virtual -Werror -Wfatal-errors -g -O0 -fno-inline -g -O2 -MT librina_la-security-manager.lo -MD -MP -MF .deps/librina_la-security-manager.Tpo -c /home/vmaffione/git/nxw/irati/librina/src/security-manager.cc  -fPIC -DPIC -o .libs/librina_la-security-manager.o
/home/vmaffione/git/nxw/irati/librina/src/security-manager.cc: In member function 'void rina::AuthSSH2PolicySet::edh_init_params()':
/home/vmaffione/git/nxw/irati/librina/src/security-manager.cc:778:15: error: invalid use of incomplete type 'DH {aka struct dh_st}'
  dh_parameters->p = BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
               ^~
compilation terminated due to -Wfatal-errors.
make[3]: *** [Makefile:1241: librina_la-security-manager.lo] Error 1
make[3]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina/src'
make[2]: *** [Makefile:1518: install] Error 2
make[2]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina/src'
make[1]: *** [Makefile:476: install-recursive] Error 1
make[1]: Leaving directory '/home/vmaffione/git/nxw/irati/build/librina'
make: *** [Makefile:773: install] Error 2
install-user-from-scratch: Cannot complete librina phase

````

--- a/README.md
+++ b/README.md
@@ -39,7 +39,7 @@ of IRATI in future releases.
 ## 2. Build instructions                                                       #
 #############################################################################
 
-### Building on Ubuntu 16.04 and Debian 8
+### Building on Ubuntu 16.04, Debian 8 and Debian 9
 #############################################################################
 
 **NOTE for Debian 8**: For the kernel modules, a Linux kernel with a version between 4.1 
@@ -49,17 +49,28 @@ and 4.9 (included) has to be installed in the system, with the kernel headers. U
 Once this is done, please install user-space dependencies
 
     $ apt-get update
-    $ apt-get install autoconf automake libtool pkg-config git g++ libssl-dev protobuf-compiler libprotobuf-dev socat
+    $ apt-get install autoconf automake libtool pkg-config git g++ libssl-dev protobuf-compiler libprotobuf-dev socat python linux-headers-$(uname -r)
     $ apt-get install hostapd (if the system will be configured as an access point)
     $ apt-get install wpasupplicant (if the system will be configured as a mobile host)
 
-If you want to generate the Java bindings, the following packages are also required
+Download the IRATI repo (arcfire branch) and enter the root directory
 
-    $ apt-get install openjdk-6-jdk maven
+    $ git clone -b arcfire https://github.com/IRATI/stack.git
+    $ cd stack
 
-Required packages to be build from source (only if generation of Java bindings is required)
+Build and install both kernel-space and user-space software
 
-* SWIG (version >= 2.0.8 required, 2.0.12 is known to be working fine) from http://swig.org
+    $ sudo ./configure --prefix <path to IRATI installation folder>
+    $ sudo make install
+
+### Building on Arch Linux
+#############################################################################
+
+Install user-space dependencies
+
+    $ pacman -S autoconf automake autoconf libtool pkg-config git openssl protobuf socat python linux-headers
+    $ pacman -S hostapd (if the system will be configured as an access point)
+    $ pacman -S wpa_supplicant (if the system will be configured as a mobile host)
 
 Download the IRATI repo (arcfire branch) and enter the root directory
 
@@ -79,7 +90,7 @@ Build and install both kernel-space and user-space software
 Insert the SD card into the Raspberry Pi and power it on. Log in with user 'pi' and 
 password 'raspberry'.  As root ("sudo su -" or equivalent):
 
-Check your kernel version (via uname -r), if it is not 4.9.24-v7+, update the distro
+Check your kernel version (via uname -r), if it is not at least 4.9.24-v7+, update the distro
 
     $ apt-get update
     $ apt-get dist-upgrade

--- a/kernel/configure
+++ b/kernel/configure
@@ -62,8 +62,8 @@ done
 
 # Check if kernel headers are there for out-of-tree build
 if [ ! -d $KERNBUILDDIR ]; then
-    printmsg "Cannot find '$KERNBUILDDIR' kernel build directory"
-    printmsg "Make sure headers for the running kernel are installed"
+    echo "Cannot find '$KERNBUILDDIR' kernel build directory"
+    echo "Make sure headers for the running kernel are installed"
     exit 255
 fi
 

--- a/librina/src/security-manager.cc
+++ b/librina/src/security-manager.cc
@@ -1364,6 +1364,8 @@ IAuthPolicySet::AuthStatus AuthSSH2PolicySet::crypto_state_updated(int port_id)
 
 IAuthPolicySet::AuthStatus AuthSSH2PolicySet::decryption_enabled_server(SSH2SecurityContext * sc)
 {
+	const BIGNUM *pub_key;
+
 	if (sc->state != SSH2SecurityContext::REQUESTED_ENABLE_DECRYPTION_SERVER) {
 		LOG_ERR("Wrong state of policy");
 		sec_man->destroy_security_context(sc->id);
@@ -1379,13 +1381,13 @@ IAuthPolicySet::AuthStatus AuthSSH2PolicySet::decryption_enabled_server(SSH2Secu
 	auth_options.encrypt_algs.push_back(sc->encrypt_alg);
 	auth_options.mac_algs.push_back(sc->mac_alg);
 	auth_options.compress_algs.push_back(sc->compress_alg);
-	auth_options.dh_public_key.data = BN_to_binary(sc->dh_state->pub_key,
-						       &auth_options.dh_public_key.length);
-
-	if (auth_options.dh_public_key.length <= 0) {
+	DH_get0_key(sc->dh_state, &pub_key, NULL);
+	auth_options.dh_public_key.length = BN_num_bytes(pub_key);
+	auth_options.dh_public_key.data = new unsigned char[auth_options.dh_public_key.length];
+	if (BN_bn2bin(pub_key, auth_options.dh_public_key.data) <= 0 ) {
 		LOG_ERR("Error transforming big number to binary");
-		sec_man->destroy_security_context(sc->id);
-		return IAuthPolicySet::FAILED;
+		delete sc;
+		throw Exception();
 	}
 
 	//Send message to peer with selected algorithms and public key

--- a/librina/src/security-manager.cc
+++ b/librina/src/security-manager.cc
@@ -875,6 +875,14 @@ void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key)
 	*pub_key = dh->pub_key;
 	*priv_key = dh->priv_key;
 }
+
+void DH_get0_pqg(const DH *dh, const BIGNUM **p,
+		 const BIGNUM **q, const BIGNUM **g)
+{
+	*p = dh->p;
+	*q = dh->q;
+	*g = dh->g;
+}
 #endif
 
 //Class AuthSSH2
@@ -1124,6 +1132,7 @@ int AuthSSH2PolicySet::load_authentication_keys(SSH2SecurityContext * sc)
 int AuthSSH2PolicySet::edh_init_keys(SSH2SecurityContext * sc)
 {
 	DH *dh_state;
+	const BIGNUM *p, *g;
 
 	// Init own parameters
 	if (!dh_parameters) {
@@ -1137,8 +1146,8 @@ int AuthSSH2PolicySet::edh_init_keys(SSH2SecurityContext * sc)
 	}
 
 	// Set P and G (re-use defaults or use the ones sent by the peer)
-	dh_state->p = BN_dup(dh_parameters->p);
-	dh_state->g = BN_dup(dh_parameters->g);
+	DH_get0_pqg(dh_parameters, &p, NULL, &g);
+	DH_set0_pqg(dh_state, BN_dup(p), NULL, BN_dup(g));
 
 	// Generate the public and private key pair
 	if (DH_generate_key(dh_state) != 1) {

--- a/librina/src/security-manager.cc
+++ b/librina/src/security-manager.cc
@@ -867,21 +867,29 @@ int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 	dh->q = q;
 	dh->g = g;
 
-	return 0;
+	return 1;
 }
 
 void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key)
 {
-	*pub_key = dh->pub_key;
-	*priv_key = dh->priv_key;
+	if (pub_key)
+		*pub_key = dh->pub_key;
+
+	if (priv_key)
+		*priv_key = dh->priv_key;
 }
 
 void DH_get0_pqg(const DH *dh, const BIGNUM **p,
 		 const BIGNUM **q, const BIGNUM **g)
 {
-	*p = dh->p;
-	*q = dh->q;
-	*g = dh->g;
+	if (p)
+		*p = dh->p;
+
+	if (q)
+		*q = dh->q;
+
+	if (g)
+		*g = dh->g;
 }
 #endif
 

--- a/librina/src/security-manager.cc
+++ b/librina/src/security-manager.cc
@@ -869,6 +869,12 @@ int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
 
 	return 0;
 }
+
+void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key)
+{
+	*pub_key = dh->pub_key;
+	*priv_key = dh->priv_key;
+}
 #endif
 
 //Class AuthSSH2
@@ -955,7 +961,7 @@ void AuthSSH2PolicySet::edh_init_params()
 		return;
 	}
 
-	if (DH_set0_pqg(dh_parameters, p, NULL, g)) {
+	if (DH_set0_pqg(dh_parameters, p, NULL, g) != 1) {
 		LOG_ERR("Problems setting P and G");
 		DH_free(dh_parameters);
 		BN_free(p);
@@ -992,6 +998,8 @@ cdap_rib::auth_policy_t AuthSSH2PolicySet::get_auth_policy(int session_id,
 							   const cdap_rib::ep_info_t& peer_ap,
 					            	   const AuthSDUProtectionProfile& profile)
 {
+	const BIGNUM *pub_key;
+
 	if (profile.authPolicy.name_ != type) {
 		LOG_ERR("Wrong policy name: %s, expected: %s",
 				profile.authPolicy.name_.c_str(),
@@ -1032,10 +1040,10 @@ cdap_rib::auth_policy_t AuthSSH2PolicySet::get_auth_policy(int session_id,
 	options.encrypt_algs.push_back(sc->encrypt_alg);
 	options.mac_algs.push_back(sc->mac_alg);
 	options.compress_algs.push_back(sc->compress_alg);
-	options.dh_public_key.data = BN_to_binary(sc->dh_state->pub_key,
-						  &options.dh_public_key.length);
-
-	if (options.dh_public_key.length <= 0 ) {
+	DH_get0_key(sc->dh_state, &pub_key, NULL);
+	options.dh_public_key.length = BN_num_bytes(pub_key);
+	options.dh_public_key.data = new unsigned char[options.dh_public_key.length];
+	if (BN_bn2bin(pub_key, options.dh_public_key.data) <= 0 ) {
 		LOG_ERR("Error transforming big number to binary");
 		delete sc;
 		throw Exception();

--- a/librina/src/security-manager.cc
+++ b/librina/src/security-manager.cc
@@ -859,6 +859,18 @@ SSH2SecurityContext::SSH2SecurityContext(int session_id,
 	state = BEGIN;
 }
 
+//OpenSSL 1.1 compatibility layer
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g)
+{
+	dh->p = p;
+	dh->q = q;
+	dh->g = g;
+
+	return 0;
+}
+#endif
+
 //Class AuthSSH2
 const int AuthSSH2PolicySet::DEFAULT_TIMEOUT = 10000;
 const std::string AuthSSH2PolicySet::EDH_EXCHANGE = "Ephemeral Diffie-Hellman exchange";
@@ -893,6 +905,7 @@ AuthSSH2PolicySet::~AuthSSH2PolicySet()
 void AuthSSH2PolicySet::edh_init_params()
 {
 	int codes;
+	BIGNUM *p, *g;
 
 	static unsigned char dh2048_p[]={
 		0xC4,0x25,0x37,0x63,0x56,0x46,0xDA,0x97,0x3A,0x51,0x98,0xA1,
@@ -927,11 +940,26 @@ void AuthSSH2PolicySet::edh_init_params()
 		return;
 	}
 
-	dh_parameters->p = BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
-	dh_parameters->g = BN_bin2bn(dh2048_g,sizeof(dh2048_g),NULL);
-	if ((dh_parameters->p == NULL) || (dh_parameters->g == NULL)) {
-		LOG_ERR("Problems converting DH parameters to big number");
+	p = BN_bin2bn(dh2048_p,sizeof(dh2048_p),NULL);
+	if (p == NULL) {
+		LOG_ERR("Problems converting P to big number");
 		DH_free(dh_parameters);
+		return;
+	}
+
+	g = BN_bin2bn(dh2048_g,sizeof(dh2048_g),NULL);
+	if (g == NULL) {
+		LOG_ERR("Problems converting G to big number");
+		DH_free(dh_parameters);
+		BN_free(p);
+		return;
+	}
+
+	if (DH_set0_pqg(dh_parameters, p, NULL, g)) {
+		LOG_ERR("Problems setting P and G");
+		DH_free(dh_parameters);
+		BN_free(p);
+		BN_free(g);
 		return;
 	}
 

--- a/rina-tools/src/key-managers/km-common.cc
+++ b/rina-tools/src/key-managers/km-common.cc
@@ -871,17 +871,28 @@ void KeyContainerManager::set_application_process(rina::ApplicationProcess * ap)
 
 int KeyContainerManager::generate_rsa_key_pair(struct key_container * kc)
 {
-	RSA *myrsa;
+	RSA * myrsa;
+	BIO * pri, * pub;
+	BIGNUM * bne = NULL;
 	unsigned long e = RSA_3;
+	int bits = 1024;
 
-	myrsa = RSA_generate_key(1024,e,NULL,NULL);
-	if (!myrsa) {
-		LOG_ERR("Problems generating RSA key");
+	bne = BN_new();
+	if (BN_set_word(bne, e) != 1) {
+		BN_free(bne);
 		return -1;
 	}
 
-	BIO *pri = BIO_new(BIO_s_mem());
-	BIO *pub = BIO_new(BIO_s_mem());
+	myrsa = RSA_new();
+	if (RSA_generate_key_ex(myrsa, bits, bne, NULL) != 1) {
+		LOG_ERR("Problems generating RSA key");
+		RSA_free(myrsa);
+		BN_free(bne);
+		return -1;
+	}
+
+	pri = BIO_new(BIO_s_mem());
+	pub = BIO_new(BIO_s_mem());
 
 	PEM_write_bio_RSAPrivateKey(pri, myrsa, NULL, NULL, 0, NULL, NULL);
 	PEM_write_bio_RSAPublicKey(pub, myrsa);
@@ -893,6 +904,9 @@ int KeyContainerManager::generate_rsa_key_pair(struct key_container * kc)
 
 	BIO_read(pri, kc->private_key.message_, kc->private_key.size_);
 	BIO_read(pub, kc->public_key.message_, kc->public_key.size_);
+
+	RSA_free(myrsa);
+	BN_free(bne);
 
 	return 0;
 }

--- a/rina-tools/src/key-managers/km-common.cc
+++ b/rina-tools/src/key-managers/km-common.cc
@@ -269,7 +269,9 @@ KMSDUProtectionHandler::KMSDUProtectionHandler()
 	/* Initialise OpenSSL library */
 	ERR_load_crypto_strings();
 	OpenSSL_add_all_algorithms();
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OPENSSL_config(NULL);
+#endif
 }
 
 void KMSDUProtectionHandler::set_security_manager(rina::ISecurityManager * sec_man)

--- a/rinad/src/ipcm/addons/ma/ribf.cc
+++ b/rinad/src/ipcm/addons/ma/ribf.cc
@@ -213,7 +213,9 @@ MASDUProtectionHandler::MASDUProtectionHandler()
 	/* Initialise OpenSSL library */
 	ERR_load_crypto_strings();
 	OpenSSL_add_all_algorithms();
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
 	OPENSSL_config(NULL);
+#endif
 }
 
 void MASDUProtectionHandler::set_security_manager(rina::ISecurityManager * sec_man)

